text/1028-cli-hooks.rst

```diff
@@ -122,23 +125,105 @@ means that the hooks associated with applying schema changes must be executed
 for these commands as well.
 
 
+Watch Configuration
+===================
+
+Sometimes in a Gel project there's a need to respond to some file changes: run
+a migration due to schema change or run code generators due to query file
+change. The CLI already supports watching the schema for changes and applying
+them, but we can generalize this mechanism and make it more flexible, much
+like the hooks that respond to CLI commands.
+
+The idea is to have a mechanism for specifying what changes you want to watch
+and a script that gets triggered by those changes. Unlike hooks, there are no
+"before" and "after" triggers here. Only one kind of response is possible:
+trigger when a change to the watched entity is detected.
+
+The watch configuration will be described in the ``gel.toml`` project manifest
+file in the ``[watch]`` table. We will introduce a ``[watch.files]`` sub-table
+for watching file system changes and triggering scripts. Potentially, we will
+also later introduce a ``[watch.gel-config]`` sub-table for monitoring changes
+to the various database config values and responding to them.
+
+The general structure of watch tables is going to use the key (potentially
+quoted) to specify *what is being watched* and the corresponding value to
+specify the script that will be triggered by changes.
+
+Since the triggered scripts will be executed in response to changes, we cannot
+meaningfully display the output of the scripts to the user. We may want to
+setup an optional logfile to record the script output. This can be specified
+in the general ``[watch]`` section as ``logfile="<path-to-logfile>"``.
+
+We may want to setup a bit of a delay before triggering the scripts so as to
+allow a sequence of changes to happen before the response script is triggered.
+This is mostly to reduce unnecessary multiple triggers for the same watched
+entity.
+
+Another consequence of executing triggered scripts in the background is that
+sometimes the changes are invalid in some way and the script will fail. This
+is considered part of normal operation (e.g. syntax error in a saved query
+file) and the scripts should be such that failure does not create some
+non-recoverable state.
+
+
+Files
+-----
+
+The ``[watch.files]`` table should contain keys that are interpreted as file system paths. The following special symbols should be allowed:
+
+- ``?`` to denote a single character wild card
+
+- ``*`` to denote an arbitrary string of characters in a name
+
+- ``**`` to denote an arbitrary sub-path
```

Comment:
I suggest expanding this to common glob syntax.

Here is a library we can use: https://docs.rs/globset/latest/globset/#syntax

Reply:
@vpetrovykh are we fine with also having `{a,b}`, `[ab]` and `[!ab]` in the glob syntax?

Reply:
Ignore this, I see that you've updated with link to the lib.